kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java

```diff
@@ -58,7 +59,41 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
                 LOGGER.debug("{}: Dispatching downstream {} request to filter{}: {}",
                         ctx.channel(), decodedFrame.apiKey(), filterDescriptor(), msg);
             }
-            invoker.onRequest(decodedFrame.apiKey(), decodedFrame.apiVersion(), decodedFrame.header(), decodedFrame.body(), filterContext);
+
+            var stage = invoker.onRequest(decodedFrame.apiKey(), decodedFrame.apiVersion(), decodedFrame.header(),
+                    decodedFrame.body(), filterContext);
+            stage.whenComplete((requestFilterResult, t) -> {
+                // maybe better to run the whole thing on the netty thread.
+
+                if (t != null) {
+                    filterContext.closeConnection();
+                    return;
+                }
+
+                if (requestFilterResult.drop()) {
+                    if (LOGGER.isDebugEnabled()) {
+                        LOGGER.debug("{}: Filter{} drops request {}",
+                                ctx.channel(), filterDescriptor(), decodedFrame.apiKey());
+                    }
+                }
+                else if (requestFilterResult.message() != null) {
+                    if (requestFilterResult.shortCircuitResponse()) {
+                        // this is the short circuit path
+                        var header = requestFilterResult.header() == null ? new ResponseHeaderData() : ((ResponseHeaderData) requestFilterResult.header());
+                        header.setCorrelationId(decodedFrame.correlationId());
+                        filterContext.forwardResponseInternal(header, requestFilterResult.message());
+                    }
+                    else {
+                        var header = requestFilterResult.header() == null ? decodedFrame.header() : requestFilterResult.header();
+                        filterContext.forwardRequestInternal((RequestHeaderData) header, requestFilterResult.message());
+                    }
+                }
```

Comment:
I'm starting to wonder if `requestFilterResult` should return an action. 

e.g. 
```java
requestFilterResult.action().execute(ctx, decodedFrame, filterContext)
```

Actual API to be confirmed...

My other alternative is to make action return an enum and we switch on that here.

Reply:
Yeah, I toyed with that idea too.  I think the idea has merit and we should keep it in mind for the next phase.

Reply:
Yeah on reflection I think that can be tackled/revisited later.